### PR TITLE
DR-3454 Snapshot/Dataset Retrieve Query Refactors

### PR DESCRIPTION
**Pull request by nmalfroy**
This PR implements the query proposal from this doc:
https://docs.google.com/document/d/1vmp81uoqS_qtswT0Jo7y76y8GRZwAdGqzRd4XiVyPP8/edit#heading=h.pjsggzrj5awt

This is to reduce load on the database.  This PR aims to reduce 1:N queries (e.g. loops that call sql queries).

While I haven't been able to really observe the load benefit in the wild, I can say that for individual requests (e.g. get dataset and get snapshot) benefits:

Before fix:
- Get dataset: ~400ms
- Get Snapshot: ~650ms

After fix:
- Get dataset: ~300ms
- Get Snapshot: ~300ms

So I expect to get some benefit just on perf in addition to taking load off of the DB which will help in general.

The areas that I focused on were:
- Combining table and column queries into one for datasets and snapshots and
- Combining the query that build mappings for snapshots into one

Oh, and I also converted some the other bigger queries in the various daos into multiline strings

### CODE REVIEW

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java [hunk 1]
```diff
@@ -27,20 +29,36 @@ public class SnapshotTableDao {
...
+  ORDER BY t.ctid, c.ordinal
+  """;
+
   private static final String sqlSelectColumn =
-      "SELECT id, name, type, array_of, required "
-          + "FROM snapshot_column "
-          + "WHERE table_id = :table_id "
-          + "ORDER BY ordinal";
+      """
+  SELECT id column_id, name column_name, type column_type, array_of column_array_of, required column_required
```
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java:57]: Let's say we add a column to any of these Postgres table definitions, like `snapshot_column` table (but this comment applies overall).

Right now, we'd have to make SQL changes in 3 places: `sqlInsertColumn`, `sqlSelectTable`, and `sqlSelectColumn`.  I could see it being easy to miss a modification.

Is there a way we can define Postgres SQL column definitions for reuse?

File: src/test/java/bio/terra/service/dataset/DatasetDaoTest.java [hunk 2]
```diff
@@ -524,23 +522,28 @@ public void primaryKeyTest() throws Exception {
 
     assertThat(
         "single-column primary keys are set correctly",
-        variants.getPrimaryKey().stream().map(Column::getName).collect(Collectors.toList()),
-        equalTo(Collections.singletonList("id")));
+        variants.getPrimaryKey(),
+        containsInAnyOrder(variants.getColumnByName("id").get()));
```
Comment (okotsopoulos) [inline src/test/java/bio/terra/service/dataset/DatasetDaoTest.java]: I think it's a good sign that these tests became more clear as a result of these changes :)

**Review by okotsopoulos — COMMENTED** [review]
Thanks for making these improvements, I have a request around maintainability but I'm happy to discuss further!

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java (hunk 1 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java:57]: I was torn around that.  I actually prefer this approach since it makes the SQL easier to read.  I don't love the pattern of building sql from concatenation.  I do think that the unit test is really meant to catch that type of mistake and I personally would take readability over the maintenance in this case the changes are so incredibly rare

Comment (okotsopoulos) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java:57]: Hm, okay, fair enough… you're right that such changes are rare!

**Review by okotsopoulos — APPROVED** [review]

File: src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java [hunk 3]
```diff
@@ -265,16 +267,10 @@ public void happyInOutTest() {
...
-            .filter(t -> t.getName().equals("thetable"))
-            .findFirst()
-            .orElseThrow(AssertionError::new);
-    Table snapshotTable2 =
-        snapshot.getTables().stream()
-            .filter(t -> t.getName().equals("anothertable"))
-            .findFirst()
-            .orElseThrow(AssertionError::new);
+    SnapshotTable snapshotTable1 =
+        fromDb.getTableByName("thetable").orElseThrow(AssertionError::new);
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java]: Why use a custom exception here? 
```suggestion
        fromDb.getTableByName("thetable").orElseThrow();
```

File: src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java [hunk 4]
```diff
@@ -265,16 +267,10 @@ public void happyInOutTest() {
             .filter(t -> t.getName().equals("thetable"))
             .findFirst()
             .orElseThrow(AssertionError::new);
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java]: Could also clean this up while you're here
```
    Table datasetTable = dataset.getTableByName("thetable").orElseThrow();
```

File: src/test/java/bio/terra/service/dataset/DatasetDaoTest.java [hunk 5]
```diff
@@ -524,23 +522,28 @@ public void primaryKeyTest() throws Exception {
...
 
     assertThat(
         "dual-column primary keys are set correctly",
-        metaAnalysis.getPrimaryKey().stream().map(Column::getName).collect(Collectors.toList()),
-        equalTo(Arrays.asList("variant_id", "phenotype")));
+        metaAnalysis.getPrimaryKey(),
+        containsInAnyOrder(
+            metaAnalysis.getColumnByName("variant_id").get(),
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/dataset/DatasetDaoTest.java]: These should probably be `getOrElseThrow()` instead of `get()`. 
```suggestion
            metaAnalysis.getColumnByName("variant_id").getOrElseThrow(),
```

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java [hunk 6]
```diff
@@ -93,42 +111,59 @@ private void createColumns(UUID tableId, Collection<Column> columns) {
   }
 
   public List<SnapshotTable> retrieveTables(UUID parentId) {
-    MapSqlParameterSource params = new MapSqlParameterSource().addValue("parent_id", parentId);
-    return jdbcTemplate.query(
+    MapSqlParameterSource params = new MapSqlParameterSource().addValue("snapshot_id", parentId);
+    Map<UUID, SnapshotTable> tableMap = new TreeMap<>();
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java:116]: Using `TreeMap` here to preserve order? A comment might be nice if that's the case.

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java [hunk 7]
```diff
@@ -93,42 +111,59 @@ private void createColumns(UUID tableId, Collection<Column> columns) {
...
+              tableMap.computeIfAbsent(
+                  tableId,
+                  (id) -> {
+                    try {
+                      return new SnapshotTable()
+                          .id(id)
+                          .name(rs.getString("table_name"))
+                          .rowCount(rs.getLong("table_row_count"))
+                          .columns(new ArrayList<>())
+                          .primaryKey(new ArrayList<>());
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java]: Since this is a type we wrote (vs one openapi generates), you could add a new constructor that does some of this work internally.
```suggestion
                      return new SnapshotTable(id, rs.getString("table_name"), rs.getLong("table_row_count"));
```

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java [hunk 8]
```diff
@@ -93,42 +111,59 @@ private void createColumns(UUID tableId, Collection<Column> columns) {
...
+                    } catch (SQLException e) {
+                      throw new RuntimeException(e);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java]: Is there no better type of exception to throw here? I would expect to see a subclass of `RuntimeException` instead.

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java [hunk 9]
```diff
@@ -93,42 +111,59 @@ private void createColumns(UUID tableId, Collection<Column> columns) {
...
+          // Add column to table
+          Column column = getColumnRowMapper(table).mapRow(rs, rowNum);
+          table.getColumns().add(column);
 
-          List<String> primaryKey = DaoUtils.getStringList(rs, "primary_key");
-          List<Column> naturalKeyColumns =
-              primaryKey.stream().map(columnMap::get).collect(Collectors.toList());
-          table.primaryKey(naturalKeyColumns);
+          // Add primary key to table if appropriate
+          List<String> primaryKey = DaoUtils.getStringList(rs, "table_primary_key");
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java]: This is the same for all rows in a table but is recomputed for each row. If you wanted to improve this, you could use another `Map.computeIfAbsent()` for `primaryKey`.

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java [hunk 10]
```diff
@@ -93,42 +111,59 @@ private void createColumns(UUID tableId, Collection<Column> columns) {
...
+          if (primaryKey.contains(column.getName())) {
+            table.getPrimaryKey().add(column);
+          }
 
           return table;
         });
+    return List.copyOf(tableMap.values());
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java:170]: It's a minor thing but we use `List` in many places where `Collection` would be fine, and this is probably one of those cases. Looking at all callers of this method, there are only five cases. Four all use `stream()` and the fifth passes it to `Collections.unmodifiableList()`. If that was changed, you could return `values()` directly.

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java [hunk 11]
```diff
@@ -93,42 +111,59 @@ private void createColumns(UUID tableId, Collection<Column> columns) {
   }
 
   public List<SnapshotTable> retrieveTables(UUID parentId) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java]: If this is the snapshot ID maybe the change the parameter name too?
```suggestion
  public List<SnapshotTable> retrieveTables(UUID snapshotId) {
```

File: src/main/java/bio/terra/service/snapshot/SnapshotTable.java [hunk 12]
```diff
@@ -66,4 +67,25 @@ public SnapshotTable primaryKey(List<Column> primaryKey) {
     this.primaryKey = primaryKey;
     return this;
   }
+
+  @Override
+  public boolean equals(Object o) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotTable.java:72]: It doesn't really make sense to have `equals()` and `hashcode()` on an object with mutable fields. Why was this change necessary?

File: src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java [hunk 13]
```diff
@@ -62,73 +65,73 @@ protected void createColumns(UUID tableId, Collection<SnapshotMapColumn> columns
...
-                  .fromColumn(datasetColumn.get())
-                  .toColumn(snapshotColumn.get());
-            });
-
-    return mapColumns;
+  private Table getTable(UUID tableId, FSContainerInterface container)
+      throws CorruptMetadataException {
+    var table =
+        switch (container.getCollectionType()) {
+          case DATASET -> ((Dataset) container).getTableById(tableId);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java:120]: Why is `getTableById()` not a method on `FSContainerInterface`? `Dataset` and `Snapshot` are the only subclasses, and they both have this method, so it seems like it wouldn't hurt to add it to the base type, and would let us clean up code like this.

File: src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java [hunk 14]
```diff
@@ -62,73 +65,73 @@ protected void createColumns(UUID tableId, Collection<SnapshotMapColumn> columns
...
   }
 
-  public List<SnapshotMapColumn> retrieveMapColumns(
-      UUID mapTableId, Table fromTable, Table toTable) {
-    String sql =
-        "SELECT id, from_column_id, to_column_id"
-            + " FROM snapshot_map_column WHERE map_table_id = :map_table_id";
-
-    List<SnapshotMapColumn> mapColumns =
-        jdbcTemplate.query(
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java:108]: I wonder if a static analysis tool can find cases like this, where we're nesting calls to `query()`. Even if we don't fix them all it'd be nice to know which ones we have and track them.

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java [hunk 15]
```diff
@@ -230,13 +275,17 @@ List<Column> retrieveColumns(Table table) {
...
-                .id(rs.getObject("id", UUID.class))
-                .table(table)
-                .name(rs.getString("name"))
-                .type(TableDataType.fromValue(rs.getString("type")))
-                .arrayOf(rs.getBoolean("array_of"))
-                .required(rs.getBoolean("required")));
+        getColumnRowMapper(table));
+  }
+
+  private RowMapper<Column> getColumnRowMapper(Table table) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: Is this the same as the one in `SnapshotTableDao`? It'd be nice to share code if possible.

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java [hunk 16]
```diff
@@ -176,45 +202,64 @@ private void removeColumns(UUID datasetId, UUID tableId, Collection<Column> colu
...
+
+          // Add column to table
+          Column column = getColumnRowMapper(table).mapRow(rs, rowNum);
+          table.getColumns().add(column);
+
+          // Add primary key to table if appropriate
+          List<String> primaryKey = DaoUtils.getStringList(rs, "table_primary_key");
+          if (primaryKey.contains(column.getName())) {
+            table.getPrimaryKey().add(column);
           }
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: It looks like this entire fragment is shared with `SnapshotTableDao`. I guess there's no good way to share code here?

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java [hunk 17]
```diff
@@ -36,23 +37,48 @@ public class DatasetTableDao {
...
+  """;
+
   private static final String sqlSelectTable =
-      "SELECT id, name, raw_table_name, soft_delete_table_name, row_metadata_table_name, primary_key, bigquery_partition_config::text, "
-          + "(bigquery_partition_config->>'version')::bigint AS bigquery_partition_config_version "
-          + "FROM dataset_table WHERE dataset_id = :dataset_id";
+      """
+  SELECT t.id table_id,
+         t.name table_name,
+         t.raw_table_name table_raw_table_name,
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/dataset/DatasetTableDao.java:58]: I guess it's nice to be consistent but only aliasing duplicate names would've also worked here and would be less verbose. It's a SQL error to have duplicate column names in a result (I believe) so you wouldn't have a case where names overlap silently.

**Review by pshapiro4broad — APPROVED** [review]
Looks OK, a few comments

File: src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java (hunk 4 above)
Comment (nmalfroy) [reply src/test/java/bio/terra/service/snapshot/SnapshotDaoTest.java]: ah right on

File: src/test/java/bio/terra/service/dataset/DatasetDaoTest.java (hunk 5 above)
Comment (nmalfroy) [reply src/test/java/bio/terra/service/dataset/DatasetDaoTest.java]: I don't think that exists...

Comment (nmalfroy) [reply src/test/java/bio/terra/service/dataset/DatasetDaoTest.java]: Oh wait, it's `orElseThrow`, nvm

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java (hunk 6 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java:116]: Yeah, and good call

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java (hunk 7 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java]: Eh, I like the fluent approach.  We talked offline about eventually moving this to a builder pattern

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java (hunk 8 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java]: Good catch!

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java (hunk 9 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java]: Yeah, I didn't know if it was worth the code complexity.  I can add it

Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java]: Actually, this is good.  I can use this to preserve the order of the primary key columns (probably not super important but might as well)

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java (hunk 10 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java:170]: The problem is that collection doesn't preserve order so we explicitly want to indicate that order _is_ deterministic in this case

File: src/main/java/bio/terra/service/snapshot/SnapshotTable.java (hunk 12 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotTable.java:72]: This is useful for testing.  This gives us a way to compare the full objects in tests

File: src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java (hunk 13 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java:120]: Ah, I tried to get it to work.  Unfortunately, they both return their own `Table` type objects and even with generic type annotations, because of type erasure, you can't set it at the interface level...it just knows that it returns an untyped `Optional`

File: src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java (hunk 14 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/snapshot/SnapshotMapTableDao.java:108]: Yeah, I'm not sure how we'd test for that

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java (hunk 16 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: Not really in a way that doesn't make the code more confusing

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java (hunk 17 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/DatasetTableDao.java:58]: Yeah, I explicitly wanted to be consistent here

File: src/main/java/bio/terra/service/dataset/DatasetTableDao.java (hunk 15 above)
Comment (nmalfroy) [reply src/main/java/bio/terra/service/dataset/DatasetTableDao.java]: Yeah, I moved this to `DaoUtils`

File: src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java (hunk 10 above)
Comment (pshapiro4broad) [reply src/main/java/bio/terra/service/snapshot/SnapshotTableDao.java:170]: > The problem is that collection doesn't preserve order

I would separate type from behavior here.

`Collection` is not a concrete type so the implementation type may or may not preserve order. If the contract is that the returned object preserves order, then it doesn't matter that the type is `Collection` vs `List`. Note that even collections like `Map` or `Set` can be order preserving even though their APIs don't guarantee it.
